Add decompositions for zero_, fill_, new_full, new_zeros, new_ones

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82332

This makes symbolic tracing tests for logsigmoid and xlogy start working again.

While I'm at it, add pin_memory and layout kwargs to empty; but they
don't actually do anything and raise an error if they are non standard.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Add decompositions for zero_, fill_, new_full, new_zeros, new_ones

This makes symbolic tracing tests for logsigmoid and xlogy start working again.

While I'm at it, add pin_memory and layout kwargs to empty; but they
don't actually do anything and raise an error if they are non standard.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
@jansel I cc'ed you because even though these decomps are done in the prevailing PrimTorch style it is not really clear if inductor would actually like this style haha

ezyang commented:
cc @zou3519 because these decomps make the composite compliance problem worse

zou3519 commented:
> cc @zou3519 because these decomps make the composite compliance problem worse

We'll figure it out eventually, but no need to block this PR on figuring out how to make them composite compliant

eellison commented on torch/_refs/__init__.py:
```diff
@@ -2998,13 +3012,60 @@ def new_empty(
+    device: Optional[torch.device] = None,
+    pin_memory: bool = False,
+) -> TensorLikeType:
+    r = a.new_empty(size, dtype=dtype, layout=layout, device=device, pin_memory=pin_memory)
```
is it worth having `new_full` as a prim operator which would cover new_zeros/new_ones/new_full without requiring two operators & mutation

ezyang replied:
Not sure but I can always route these to `new_full`, at least...

eellison reviewed (approved):
LGTM

ezyang committed:
Update on "Add decompositions for zero_, fill_, new_full, new_zeros, new_ones"


This makes symbolic tracing tests for logsigmoid and xlogy start working again.

While I'm at it, add pin_memory and layout kwargs to empty; but they
don't actually do anything and raise an error if they are non standard.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add decompositions for zero_, fill_, new_full, new_zeros, new_ones"


This makes symbolic tracing tests for logsigmoid and xlogy start working again.

While I'm at it, add pin_memory and layout kwargs to empty; but they
don't actually do anything and raise an error if they are non standard.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -g

ezyang commented:
@pytorchbot merge -f "known master failure"
